cudax/include/cuda/experimental/__async/basic_sender.cuh

```diff
@@ -0,0 +1,242 @@
+
+template <class Completions>
+constexpr bool _has_stopped =
+  !_CUDA_VSTD::is_same_v<__async::completion_signatures<>,
+                         __async::transform_completion_signatures<Completions,
+                                                                  __async::completion_signatures<>,
+                                                                  _ignore_value_signature,
+                                                                  _ignore_error_signature>>;
+
+template <bool PotentiallyThrowing, class Rcvr>
+void set_current_exception_if(Rcvr& rcvr) noexcept
+{
+  if constexpr (PotentiallyThrowing)
+  {
+    __async::set_error(_CUDA_VSTD::move(rcvr), ::std::current_exception());
+  }
+}
+
+// A generic type that holds the data for an async operation, and
+// that provides a `start` method for enqueuing the work.
+template <class Sndr, class Data, class Rcvr>
+struct basic_opstate
+{
+  using _rcvr_t        = basic_receiver<Data, Rcvr>;
+  using _completions_t = completion_signatures_of_t<Sndr, _rcvr_t>;
+  using _traits_t      = _mk_completions<_has_stopped<_completions_t>, Data, Rcvr>;
+
+  using completion_signatures =
+    transform_completion_signatures<_completions_t,
+                                    __async::completion_signatures<>, // TODO
+                                    _traits_t::template _set_value_t,
+                                    _traits_t::template _set_error_t,
+                                    typename _traits_t::_set_stopped_t>;
+
+  _CCCL_HOST_DEVICE basic_opstate(Sndr&& sndr, Data data, Rcvr rcvr)
+      : state_{static_cast<Data&&>(data), static_cast<Rcvr&&>(rcvr)}
+      , op_(__async::connect(static_cast<Sndr&&>(sndr), _rcvr_t{state_}))
+  {}
+
+  _CCCL_HOST_DEVICE _CUDAX_ALWAYS_INLINE void start() noexcept
+  {
+    __async::start(op_);
+  }
+
+  state<Data, Rcvr> state_;
+  __async::connect_result_t<Sndr, _rcvr_t> op_;
+};
+
+template <class Sndr, class Rcvr>
+_CCCL_HOST_DEVICE _CUDAX_ALWAYS_INLINE auto _make_opstate(Sndr sndr, Rcvr rcvr)
+{
+  auto [tag, data, child] = _CUDA_VSTD::move(sndr);
+  return basic_opstate(_CUDA_VSTD::move(child), _CUDA_VSTD::move(data), _CUDA_VSTD::move(rcvr));
+}
+
+template <class Data, class... Sndrs>
+_CCCL_HOST_DEVICE _CUDAX_ALWAYS_INLINE auto
+_get_attrs(int, const Data& data, const Sndrs&... sndrs) noexcept -> decltype(data.get_attrs(sndrs...))
+{
+  return data.get_attrs(sndrs...);
+}
+
+template <class Data, class... Sndrs>
+_CCCL_HOST_DEVICE _CUDAX_ALWAYS_INLINE auto
+_get_attrs(long, const Data& data, const Sndrs&... sndrs) noexcept -> decltype(__async::get_env(sndrs...))
+{
+  return __async::get_env(sndrs...);
+}
+
+template <class Data, class... Sndrs>
+struct basic_sender;
+
+template <class Data, class Sndr>
+struct basic_sender<Data, Sndr>
+{
+  using sender_concept = __async::sender_t;
+  using _tag_t         = typename Data::sender_tag;
+  using _rcvr_t        = typename Data::receiver_tag;
+
+  [[no_unique_address]] _tag_t _tag_;
```

Comment:
This needs to be conditional on compiler support.
```suggestion
  _CCCL_NO_UNIQUE_ADDRESS _tag_t _tag_;
```
Also because MSVC warns about this attribute prior to C++20 we also need
```cpp
_CCCL_DIAG_PUSH
_CCCL_DIAG_SUPPRESS_MSVC(4848) // [[no_unique_address]] prior to C++20 as a vendor extension
...
_CCCL_DIAG_POP
```

Somewhere around the usage points

Reply:
i can suppress this warning in `prologue.cuh` and retore it in `epilogue.cuh`.